src/Symfony/Component/Uid/BinaryUtil.php

```diff
@@ -126,4 +126,19 @@ public static function timeToFloat(string $time): float
 
         return self::toBase($time, self::BASE10) / 10000000;
     }
+
+    /**
+     * @param float $time Seconds since the Unix epoch 1970-01-01 00:00:00
+     *
+     * @return string Count of 100-nanosecond intervals since the UUID epoch 1582-10-15 00:00:00 in hexadecimal
+     */
+    public static function floatToTime(float $time): string
+    {
+        if (\PHP_INT_SIZE >= 8) {
+            return str_pad(dechex(self::TIME_OFFSET_INT + (int) round($time * 10000000)), 16, '0', \STR_PAD_LEFT);
+        }
+
+        // TODO
```

Comment:
@nicolas-grekas I need you to do this part please

Reply:
Here is the implementation if I'm not wrong. This fails on the test with a negative timestamp, but I think this test is wrong. Neither UUID nor ULID use signed timestamps AFAIK.

```diff
--- a/src/Symfony/Component/Uid/BinaryUtil.php
+++ b/src/Symfony/Component/Uid/BinaryUtil.php
@@ -40,6 +40,7 @@ class BinaryUtil
     // 0x01b21dd213814000 is the number of 100-ns intervals between the
     // UUID epoch 1582-10-15 00:00:00 and the Unix epoch 1970-01-01 00:00:00.
     private const TIME_OFFSET_INT = 0x01b21dd213814000;
+    private const TIME_OFFSET_BIN = "\x01\xb2\x1d\xd2\x13\x81\x40\x00";
     private const TIME_OFFSET_COM = "\xfe\x4d\xe2\x2d\xec\x7e\xc0\x00";
 
     public static function toBase(string $bytes, array $map): string
@@ -135,10 +136,13 @@ class BinaryUtil
     public static function floatToTime(float $time): string
     {
         if (\PHP_INT_SIZE >= 8) {
-            return str_pad(dechex(self::TIME_OFFSET_INT + (int) round($time * 10000000)), 16, '0', \STR_PAD_LEFT);
+            return str_pad(dechex(self::TIME_OFFSET_INT + 10000000 * $time), 16, '0', \STR_PAD_LEFT);
         }
 
-        // TODO
-        return 'foo';
+        $time = sprintf('%.0F', 10000000 * $time);
+        $time = self::fromBase($time, self::BASE10);
+        $time = str_pad($time, 8, "\0", \STR_PAD_LEFT);
+
+        return bin2hex(self::add($time, self::TIME_OFFSET_BIN));
     }
 }
```

Reply:
They don't use signed timestamps but we chose to use timestamps based on the unix epoch for both uuid and ulid. Therefore, generating an uuid for a date in 1960 requires a negative timestamp. Ulid are not concerned.

Reply:
Got it, PR updated.